compaction/compaction_manager.cc

```diff
@@ -351,15 +362,12 @@ future<sstables::compaction_result> compaction_task_executor::compact_sstables(s
         auto sst = t.make_sstable();
         return sst;
     };
-    descriptor.replacer = [this, &t, release_exhausted] (sstables::compaction_completion_desc desc) {
+    descriptor.replacer = [this, &t, on_replace=std::move(on_replace)] (sstables::compaction_completion_desc desc) {
         t.get_compaction_strategy().notify_completion(t, desc.old_sstables, desc.new_sstables);
         _cm.propagate_replacement(t, desc.old_sstables, desc.new_sstables);
-        auto old_sstables = desc.old_sstables;
+        assert(on_replace);
+        on_replace(desc);
         t.on_compaction_completion(std::move(desc), sstables::offstrategy::no).get();
-        // Calls compaction manager's task for this compaction to release reference to exhausted SSTables.
```

Comment:
there's a reason why release_exhausted() comes after on_compaciton_completion, and the reason is that we should only deregister input sstables after we have removed them from table's sstable set. otherwise we risk regular compaction picking them once they were deregistered but not removed by on_compaction_completion() call that may defer before completing the removal. I'd suggest documenting it once we're here.

Reply:
sure, i added a comment to explain this.